funsor/montecarlo.py

```diff
@@ -2,57 +2,34 @@
 # SPDX-License-Identifier: Apache-2.0
 
 from collections import OrderedDict
-from contextlib import contextmanager
 
 from funsor.integrate import Integrate
-from funsor.interpreter import dispatched_interpretation, interpretation
-from funsor.terms import Funsor, eager
+from funsor.interpreter import StatefulInterpretation
+from funsor.terms import Funsor
 
 
-@dispatched_interpretation
-def monte_carlo(cls, *args):
+class MonteCarlo(StatefulInterpretation):
     """
     A Monte Carlo interpretation of :class:`~funsor.integrate.Integrate`
-    expressions. This falls back to :class:`~funsor.terms.eager` in other
-    cases.
-    """
-    # TODO Memoize sample statements in a context manager.
-    result = monte_carlo.dispatch(cls, *args)(*args)
-    if result is None:
-        result = eager(cls, *args)
-    return result
-
+    expressions. This falls back to the previous interpreter in other cases.
 
-# This is a globally configurable parameter to draw multiple samples.
-monte_carlo.sample_inputs = OrderedDict()
+    :param rng_key:
+    """
+    def __init__(self, *, rng_key=None, **sample_inputs):
+        self.rng_key = rng_key
+        self.sample_inputs = OrderedDict(sample_inputs)
 
 
-@contextmanager
-def monte_carlo_interpretation(**sample_inputs):
-    """
-    Context manager to set ``monte_carlo.sample_inputs`` and
-    install the :func:`monte_carlo` interpretation.
-    """
-    old = monte_carlo.sample_inputs
-    monte_carlo.sample_inputs = OrderedDict(sample_inputs)
-    try:
-        with interpretation(monte_carlo):
-            yield
-    finally:
-        monte_carlo.sample_inputs = old
-
-
-@monte_carlo.register(Integrate, Funsor, Funsor, frozenset)
-def monte_carlo_integrate(log_measure, integrand, reduced_vars):
-    # FIXME: how to pass rng_key to here?
-    sample = log_measure.sample(reduced_vars, monte_carlo.sample_inputs)
+@MonteCarlo.register(Integrate, Funsor, Funsor, frozenset)
+def monte_carlo_integrate(state, log_measure, integrand, reduced_vars):
+    # FIXME use state.rng_key to here
```

Comment:
@fehiepsi can you help me out with the `rng_key` logic here? I have removed the `xfail` from your test case in hopes that we can fix the issue in this PR.

Reply:
Do we only use `monte_carlo_integrate` one? If so, I think you can simply add `rng_key=state.rng_key` keyword:
```
sample = log_measure.sample(reduced_vars, state.sample_inputs, rng_key=state.rng_key)
```

But if this code is run multiple time, then we need to split the key:
```python
from funsor.util import get_backend
...
    if state.rng_key is not None and get_backend() == "jax":
        import jax

        rng_key, state.rng_key = jax.random.split(state.rng_key)
    sample = log_measure.sample(reduced_vars, state.sample_inputs, rng_key=rng_key)
```

Reply:
@fritzo Could you help me replace `if rng_key is not None` at [this line](https://github.com/pyro-ppl/funsor/blob/master/funsor/cnf.py#L89) and 10 lines below by `if rng_key is not None and get_backend() == "jax"`? I feel doing so is safer.

Reply:
Thanks for the help, jax tests now pass :tada: